### PR TITLE
Powerloop Implementation

### LINKED ISSUE #2597: Powerloop firmware to perform control commands
### Description of the task

- Call associated functions given power control msg
- Setup interrupts to read values back

### Acceptance criteria

- [ ]

### LINKED ISSUE #2646: Breakbeam Interrupt firmware

### LINKED ISSUE #2647: Power monitor firmware

### PR DESCRIPTION
**Pull request by LiCody**
### Description

- Adds classes for powerboard components
- Adds executor for power control proto
- lot of statics due to interrupts only being able to call static void(void) functions

### Testing Done

Tested Uart communication with power board

### Resolved Issues

resolves #2597 
resolves #2647
resolves #2646 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/power/charger.cpp
Comment (akhilveeraghanta) [inline src/software/power/charger.cpp]: there might be some additional conversion needed here to get the voltage from the ADC reading

@ds570001 would be a good person to talk to

File: src/software/power/charger.cpp
Comment (akhilveeraghanta) [inline src/software/power/charger.cpp:20]: if there is a `flyback_fault` we shouldn't run the charge_done_callback

or we should pass the fault into the `charge_done_callback` and have the callback handle it

File: src/software/power/charger.cpp
Comment (akhilveeraghanta) [inline src/software/power/charger.cpp]: nit: it would be good if we had `Charger::chargeCapaciators` and `Charger::dischargeCapacitors`

just to be a bit more clear vs passing in high/log as an int

File: src/software/power/chicker.h
Comment (akhilveeraghanta) [inline src/software/power/chicker.h]: nit: these can also be camel case

File: src/software/power/geneva.cpp
Comment (akhilveeraghanta) [inline src/software/power/geneva.cpp]: nit: we can remove this, itll be in the git history :)

File: src/software/power/control_executor.cpp
Comment (akhilveeraghanta) [inline src/software/power/control_executor.cpp:17]: this line is duplicated in all cases, can you call it once at the top?

File: src/software/power/control_executor.cpp
Comment (akhilveeraghanta) [inline src/software/power/control_executor.cpp]: should u also call the geneva done callback here? and move this out before setChargeMode?

File: src/software/power/control_executor.cpp
Comment (akhilveeraghanta) [inline src/software/power/control_executor.cpp]: we don't need to set the geneva for chipping, it can stay where it is

File: src/software/power/control_executor.cpp
Comment (akhilveeraghanta) [inline src/software/power/control_executor.cpp]: see dannons comment here: https://discord.com/channels/711281059857432668/764565540819042342/979489382681706596

we should set the charge mode HIGH, and then have the kick callback, when the charge done pin goes high

File: src/software/power/geneva.cpp
Comment (akhilveeraghanta) [inline src/software/power/geneva.cpp]: can you tag #2458 with the todos throughout this file

File: src/software/power/control_executor.cpp
Comment (akhilveeraghanta) [inline src/software/power/control_executor.cpp]: ditto above, we don't need to move geneva for chipping

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
nice work and nice abstractions! :battery: :zap:

File: src/software/power/chicker.cpp
Comment (akhilveeraghanta) [inline src/software/power/chicker.cpp]: TODO #2645

File: src/software/power/geneva.cpp
Comment (amrm3lm) [inline src/software/power/geneva.cpp]: attach to ticket

File: src/software/power/geneva.cpp
Comment (amrm3lm) [inline src/software/power/geneva.cpp]: attach to ticket

File: src/software/power/charger.h
Comment (amrm3lm) [inline src/software/power/charger.h]: what is `IRAM_ATTR` ?

**Review by amrm3lm — COMMENTED** [review]
very readable code, no major comments. i wonder if we need to bring back the `firmware/board` file hierarchy, because i dont feel like this belongs under `software`

File: src/software/power/control_executor.cpp
Comment (LiCody) [reply src/software/power/control_executor.cpp:17]: I dont think we want it triggered in the default case when a chicker command isn't specified

File: src/software/power/charger.h
Comment (LiCody) [reply src/software/power/charger.h]: It tells the compiler to put the function into embedded ram so we can call the function faster than if it was stored in flash and since chargeDone is run as an interrupt service routine we want to run it as fast as possible.  More info: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/memory-types.html#iram-instruction-ram

File: src/software/power/control_executor.cpp
Comment (LiCody) [reply src/software/power/control_executor.cpp]: yea thats currently what it does we register a callback (kick/chip) that gets triggered when charge done pin goes from low->high. So we want to have a delay between charges? that seems like it would affect gameplay

Comment (akhilveeraghanta) [reply src/software/power/control_executor.cpp]: We need to explicitly set chargeMode low after we kick I think, lets check with @ds570001 for the exact sequence should be/what we can't do

File: src/software/power/charger.cpp
Comment (ds570001) [reply src/software/power/charger.cpp]: The ADC resolution that the esp reads is 12 bit. This means the max value analogread() will output is 2^12 = 4096, which would correspond to a voltage of 3.2-3.3V (ESP32 has trouble distinguishing between 3.2-3.3V and 0-0.1V - not a problem for us tho because the max voltage we should read is about 3V when caps are charged to 238V). When you call analogread, it will give a value between 0 and 4096, and to get the actual voltage you need to do the following: Actual voltage = output from AnalogRead() / 4096 * 3.3V

File: src/software/power/control_executor.cpp
Comment (ds570001) [reply src/software/power/control_executor.cpp]: At the moment I haven’t tried putting the flyback in regulation mode, so to keep the capacitors charged we will need to toggle the charge pin every so often if we want to keep the capacitors constantly charged at 238V. I think if we toggle it ~ every 5s then it will stay maximally charged. if we toggle it too quickly the capacitors will overcharge and could potentially break. I will make this a priority to test out though, because if I can put the chip in regulation mode then you just need to hold charge high and the voltage will stay at 238V. However it would be nice to have this 5s toggle functionality implemented in case anything goes wrong with the regulation. If the regulation works, CHARGE just needs to be held high

File: src/software/power/charger.cpp
Comment (akhilveeraghanta) [inline src/software/power/charger.cpp]: see: https://discord.com/channels/711281059857432668/764565540819042342/980927731527208970
done pin goes low when its done charging, this should be FALLING

File: src/software/power/geneva.h
Comment (itsarune) [inline src/software/power/geneva.h]: nit: "amount"

File: src/software/power/pins.h
Comment (itsarune) [inline src/software/power/pins.h:3]: I don't think this constants file needs an Arduino header

**Review by itsarune — COMMENTED** [review]
nice work! just some nits

File: src/software/power/pins.h
Comment (LiCody) [reply src/software/power/pins.h:3]: the header contains uint8_t the header is already included in other files so there shouldn't be any difference between this and <cstdint>

**Review by itsarune — DISMISSED** [review]
great work!

**Comment by akhilveeraghanta** [comment]
CLANG_FORMAT_EXTENSIONS=(h cpp c cc hpp tpp proto)
 
can you change line 19 in fix_formatting.sh to the above, some of the powerloop files aren't being formatted correctly because they have a cc extension

**Review by itsarune — COMMENTED** [review]
why does powerloop_main have a .cc file extension?

I feel that we should just stick to the standard .h/.cpp standard we have for the rest of the codebase

**Comment by liambontkes** [comment]
Would further electrical testing be out of the scope? I think it would be good to validate the rest of the functionality in the electrical stack-up before merging (and also update this in **Testing Done**).

**Comment by LiCody** [comment]
> why does powerloop_main have a .cc file extension?
> 
> I feel that we should just stick to the standard .h/.cpp standard we have for the rest of the codebase

A .cc file is required by the bazel rule

File: src/software/power/pins.h
Comment (akhilveeraghanta) [inline src/software/power/pins.h:36]: can we use the value already in src/shared/constants.h?

**Review by akhilveeraghanta — APPROVED** [review]
lgtm! :battery: 

just one non-blocking comment

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
was testing a bit more, it looks like we are missing a piece of code in thunderloop to run the power service

can we include that in this PR?

**Comment by akhilveeraghanta** [comment]
i tried adding a prelimary poll and the loops getting stuck

File: src/software/power/pins.h
Comment (LiCody) [reply src/software/power/pins.h:36]: not currently if we update constants to be constexpr then we could

**Comment by LiCody** [comment]
> i tried adding a prelimary poll and the loops getting stuck

Are you getting anything if you open a serial monitor/screen?

Might need to run these commands to enable uart
```
systemctl stop nvgetty
systemctl disable nvgetty
udevadm trigger
sudo adduser robot dialout
```

**Review by akhilveeraghanta — APPROVED** [review]
discussed offline: won't be blocking this PR on thunderloop integration, that will come in another PR

**Review by itsarune — APPROVED** [review]

**Pull request merged**
